File: src/main/java/com/raychenon/leetcode/ProductofArrayExceptSelf.java
package com.raychenon.leetcode;

/**
 * User: raychenon
 * Date: 2019-08-18
 * https://leetcode.com/problems/product-of-array-except-self/
 */
public class ProductofArrayExceptSelf {

    public int[] productExceptSelf(int[] nums) {
        int size = nums.length;
        // quote : Given an array nums of n integers where n > 1
        if (size <= 0) {
            return new int[0];
        }

        int output[] = new int[size];

        //  prodLeft  i   prodRight
        //  ---------> <-----------
        // [ | | | | | | | | | | | ]
        int prodLeft[] = new int[size];
        int prodRight[] = new int[size];

        prodLeft[0] = 1;
        for (int i = 1; i < size; i++) {
            prodLeft[i] = prodLeft[i - 1] * nums[i - 1];
        }

        prodRight[size - 1] = 1;
        for (int i = size - 2; i >= 0; i--) {
            prodRight[i] = prodRight[i + 1] * nums[i + 1];
        }

        for (int i = 0; i < size; i++) {
            int prod = prodLeft[i] * prodRight[i];
            output[i] = prod;
        }

        return output;
    }

    public int[] productExceptSelfConstantSpace(int[] nums) {
        int size = nums.length;
        // quote : Given an array nums of n integers where n > 1
        if (size <= 0) {
            return new int[0];
        }

        int output[] = new int[size];

        //  prodLeft  i
        //  ---------> |-----------
        // [ | | | | | | | | | | | ]
        output[0] = 1;
        for (int i = 1; i < size; i++) {
            output[i] = output[i - 1] * nums[i - 1];
        }

        //            i   Right
        //  -------- |  <-----------
        // [ | | | | | | | | | | | ]
        int prodRight = 1;
        for (int i = size - 1; i >= 0; i--) {
            output[i] = output[i] * prodRight;
            prodRight = prodRight * nums[i];
        }

        return output;
    }
}

File: src/main/kotlin/com/raychenon/kotlin/leetcode/ProductofArrayExceptSelf.kt
package com.raychenon.kotlin.leetcode

/**
 * User: raychenon
 * Date: 2019-08-19
 * https://leetcode.com/problems/product-of-array-except-self/
 */
object ProductofArrayExceptSelf {

    fun productExceptSelf(nums: IntArray): IntArray {
        if (nums.isEmpty()) {
            return intArrayOf()
        }
        val size = nums.size
        val output = IntArray(size)

        //  prodLeft  i   prodRight
        //  ---------> <-----------
        // [ | | | | | | | | | | | ]
        val leftProd = IntArray(size)
        val rightProd = IntArray(size)

        rightProd[size - 1] = 1
        for (i in (size - 2) downTo 0) {
            rightProd[i] = rightProd.get(i + 1) * nums.get(i + 1)
        }

        leftProd[0] = 1
        for (i in 1 until size) {
            leftProd[i] = leftProd.get(i - 1) * nums.get(i - 1)
        }

        for (i in 0 until size) {
            output[i] = leftProd.get(i) * rightProd.get(i)
        }

        return output
    }


    /**
     * time complexity: O(n)
     * space complexity: O(1)
     */
    fun productExceptSelfConstantSpace(nums: IntArray): IntArray {
        if (nums.isEmpty()) {
            return intArrayOf()
        }
        val size = nums.size
        val output = IntArray(size)

        //  prodLeft  i   prodRight
        //  ---------> <-----------
        // [ | | | | | | | | | | | ]
        output[0] = 1
        for (i in 1 until size) {
            output[i] = output.get(i - 1) * nums.get(i - 1)
        }

        var R = 1
        for (i in (size - 1) downTo 0) {
            output[i] = output.get(i) * R
            R = R * nums.get(i)
        }

        return output
    }
}
